fragment.go

```diff
@@ -1323,73 +1323,67 @@ func (f *fragment) mergeBlock(id int, data []pairSet) (sets, clears []pairSet, e
 }
 
 // bulkImport bulk imports a set of bits and then snapshots the storage.
-// This does not affect the fragment's cache.
+// The cache is updated to reflect the new data.
 func (f *fragment) bulkImport(rowIDs, columnIDs []uint64) error {
-	f.mu.Lock()
-	defer f.mu.Unlock()
 	// Verify that there are an equal number of row ids and column ids.
 	if len(rowIDs) != len(columnIDs) {
 		return fmt.Errorf("mismatch of row/column len: %d != %d", len(rowIDs), len(columnIDs))
 	}
 
 	// Disconnect op writer so we don't append updates.
-	f.storage.OpWriter = nil
-
+	localBitmap := roaring.NewBitmap()
+	localBitmap.OpWriter = nil
 	// Process every bit.
 	// If an error occurs then reopen the storage.
 	lastID := uint64(0)
-	if err := func() error {
-		set := make(map[uint64]struct{})
-		for i := range rowIDs {
-			rowID, columnID := rowIDs[i], columnIDs[i]
+	set := make(map[uint64]struct{})
+	for i := range rowIDs {
+		rowID, columnID := rowIDs[i], columnIDs[i]
 
-			// Determine the position of the bit in the storage.
-			pos, err := f.pos(rowID, columnID)
-			if err != nil {
-				return errors.Wrap(err, "getting bit pos")
-			}
-
-			// Write to storage.
-			_, err = f.storage.Add(pos)
-			if err != nil {
-				return errors.Wrap(err, "writing")
-			}
-			// Reduce the StatsD rate for high volume stats
-			f.stats.Count("ImportBit", 1, 0.0001)
-			// import optimization to avoid linear foreach calls
-			// slight risk of concurrent cache counter being off but
-			// no real danger
-			if i == 0 || rowID != lastID {
-				lastID = rowID
-				set[rowID] = struct{}{}
-			}
-
-			// Invalidate block checksum.
-			delete(f.checksums, int(rowID/HashBlockSize))
+		// Determine the position of the bit in the storage.
+		pos, err := f.pos(rowID, columnID)
+		if err != nil {
+			return err
 		}
 
-		// Update cache counts for all rows.
-		for rowID := range set {
-			// Import should ALWAYS have row() load a new row from fragment.storage
-			// because the row that's in rowCache hasn't been updated with
-			// this import's data.
-			f.cache.BulkAdd(rowID, f.unprotectedRow(rowID, false, false).Count())
+		// Write to storage.
+		_, err = localBitmap.Add(pos)
+		if err != nil {
+			return err
+		}
+		// Reduce the StatsD rate for high volume stats
+		f.stats.Count("ImportBit", 1, 0.0001)
+		// import optimization to avoid linear foreach calls
+		// slight risk of concurrent cache counter being off but
+		// no real danger
+		if i == 0 || rowID != lastID {
+			lastID = rowID
+			set[rowID] = struct{}{}
 		}
 
-		f.cache.Invalidate()
-		return nil
-	}(); err != nil {
-		_ = f.closeStorage()
-		_ = f.openStorage()
-		return err
+		// Invalidate block checksum.
+		delete(f.checksums, int(rowID/HashBlockSize))
 	}
 
-	// Write the storage to disk and reload.
-	if err := f.snapshot(); err != nil {
-		return errors.Wrap(err, "snapshotting")
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	//f.storage.Unmmap()
+	// Update cache counts for all rows.
```

Comment:
It seems like this comment should be something like "Merge localBitmap into existing storage."